File: DepthFirstSearch/DFSOnTree/Python/MaxDepth.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def dfsHelper(root: Node) -> int:
    #individual node does not add to the height. This accommodates for that.
    leftHeight = 0
    rightHeight = 0
    if(root.right is None and root.left is None):
        return 0
    else:
        if(root.left is not None):
            leftHeight = 1 + dfsHelper(root.left)
        if(root.right is not None):
            rightHeight = 1 + dfsHelper(root.right)
    return max(leftHeight, rightHeight)

def tree_max_depth(root: Node) -> int:
    #Need to handle special case of null root
    if root is not None:
        return dfsHelper(root)
    else:
        return 0

File: DepthFirstSearch/DFSOnTree/C#/MaxDepth.cs
/*
Max depth of a binary tree is the longest root-to-leaf path. Given a binary tree, find its max depth. Here, we define the length of the path to be the number of edges on that path, not the number of nodes.
*/

using System;
using System.Collections.Generic;
using System.Linq;

class Solution
{
    public class Node<T>
    {
        public T val;
        public Node<T> left;
        public Node<T> right;

        public Node(T val)
        {
            this.val = val;
        }

        public Node(T val, Node<T> left, Node<T> right)
        {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public static int TreeMaxDepth(Node<int> root)
    {
        if(root is null){
            return 0;
        }
        else{
            int left = -1;
            int right = -1;
            if(root.left is not null){
                left = TreeMaxDepth(root.left);
            }
            if(root.right is not null){
                right = TreeMaxDepth(root.right);
            }
            return Math.Max(left, right) + 1;
        }
    }
}
